api/client/joinservice_test.go

```diff
@@ -0,0 +1,110 @@
+package client
+
+import (
+	"context"
+	"net"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+)
+
+type mockJoinServiceServer struct {
+	*proto.UnimplementedJoinServiceServer
+	registerUsingTPMMethod func(srv proto.JoinService_RegisterUsingTPMMethodServer) error
+}
+
+func (m *mockJoinServiceServer) RegisterUsingTPMMethod(srv proto.JoinService_RegisterUsingTPMMethodServer) error {
+	return m.registerUsingTPMMethod(srv)
+}
+
+func TestJoinServiceClient_RegisterUsingTPMMethod(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	lis := bufconn.Listen(100)
+	t.Cleanup(func() {
+		assert.NoError(t, lis.Close())
+	})
+
+	mockInitReq := &proto.RegisterUsingTPMMethodInitialRequest{
+		JoinRequest: &types.RegisterUsingTokenRequest{
+			Token: "token",
+		},
+	}
+	mockChallenge := &proto.TPMEncryptedCredential{
+		CredentialBlob: []byte("cred-blob"),
+		Secret:         []byte("secret"),
+	}
+	mockChallengeResp := &proto.RegisterUsingTPMMethodChallengeResponse{
+		Solution: []byte("solution"),
+	}
+	mockCerts := &proto.Certs{
+		TLS: []byte("cert"),
+	}
+	mockService := &mockJoinServiceServer{
+		registerUsingTPMMethod: func(srv proto.JoinService_RegisterUsingTPMMethodServer) error {
+			req, err := srv.Recv()
+			require.NoError(t, err)
```

Comment:
This is effectively a t.FailNow in a distinct goroutine, which [technically isn't allowed][1].

Same for all requires in this block/func.

[1]: https://pkg.go.dev/testing#T.FailNow

Reply:
[3eec5a1](https://github.com/gravitational/teleport/pull/40647/commits/3eec5a14e4903508e7ceb93839721b14230a7448)